x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -542,31 +517,97 @@ private static MvExpand descendantMvExpand(UnaryPlan unary) {
-            UnaryPlan plan = mvExpand;
-            while (plan instanceof Aggregate == false) {
-                if (plan instanceof Limit limit) {
-                    return limit;
+    protected static class PushDownMvExpandPastProject extends OptimizerRules.OptimizerRule<MvExpand> {
+        @Override
+        protected LogicalPlan rule(MvExpand mvExpand) {
+            LogicalPlan child = mvExpand.child();
+            if (child instanceof Project) {
+                return pushDownPastProject(mvExpand);
+            }
+
+            return mvExpand;
+        }
+    }
+
+    static class DuplicateLimitAfterMvExpand extends OptimizerRules.OptimizerRule<Limit> {
+        @Override
+        protected LogicalPlan rule(Limit limit) {
+            var child = limit.child();
+            // let the PushDownAndCombineLimits rule to push Limit further down
+            var shouldSkip = isTransparentForLimit(child) || child instanceof Limit;
+            if (shouldSkip || child instanceof UnaryPlan == false) {
+                return limit;
+            }
+
+            UnaryPlan unary = (UnaryPlan) child;
+            MvExpand mvExpand = null;
+            AttributeSet filterReferences = new AttributeSet();
+            boolean hasOrderBy = false;
+            boolean isFiltered = false;
+            // look for any orderBy between this limit and mv_expand
+            // and, separately, for any filter that is filtering the expanded values of mv_expand
+            while (mvExpand == null && unary instanceof UnaryPlan) {
+                if (unary instanceof MvExpand mve) {
+                    if (filterReferences.isEmpty() == false) {
+                        if (filterReferences.contains(mve.expanded())) {// if the filter is on the mv_expand field
+                            isFiltered = true;
+                        }
+                    }
+                    mvExpand = mve;
+                    break;
+                } else if (unary instanceof Filter filter) {
+                    // gather all the filters' references to be checked later when a mv_expand is found
+                    filterReferences.addAll(filter.references());
+                } else if (unary instanceof OrderBy) {
+                    hasOrderBy = true;
+                    break;
+                } else if (unary instanceof Limit) {
+                    break;
                 }
-                if (plan.child() instanceof UnaryPlan unaryPlan) {
-                    plan = unaryPlan;
+
+                if (unary.child() instanceof UnaryPlan unaryPlan) {
+                    unary = unaryPlan;
                 } else {
                     break;
                 }
             }
-            return null;
+
+            if (mvExpand != null) {
+                // if the filter is on the expanded values and the order of the results doesn't change, push down mv_expand past sort
+                if (isFiltered && hasOrderBy == false) {
```

Comment:
I think `mvExpand == null` and `hasOrderBy` are mutually exclusive. If so, maybe this can be written a bit easier.